xt_training/utils/testing.py

```diff
@@ -49,13 +49,7 @@ def test(args):
         on_exit
     )
 
-    # Save config file
-    try:
-        if isinstance(config_path, str):
-            shutil.copy(config_path, f'{save_dir}/config.py')
-        else:
-            shutil.copy(config_path['__file__'], f'{save_dir}/config.py')
-    except shutil.SameFileError:
-        pass
+    # Save config file and repo state in checkpoint directory
+    _save_state(save_dir, config_path)
```

Comment:
Hi Tim, very nice work. Just wanted to check: IIUC you've only added this to the Command Line API so far. Is there an argument for making this part of the Functional API?

Reply:
Never mind, I see now. We don't have a way of saving the config in the Functional API case.

Reply:
This is still a good point however. It would be nice if we had a way to do this even for interactive python sessions. A few ideas:

* We could check if python is currently running in an interactive terminal with `hasattr(sys, 'ps1')` (`True` when interactive - works for IPython, Jupyter, and python built-in terminal)
* For IPython and Jupyter (just IPython behind-the-scenes) we can actually extract the full set of commands that have been run in the current python session with:
    ```python
    with open(os.path.join(save_dir, 'config.py'), 'w') as f:
        f.write('\n'.join(In))
    ```
    This is correct even when cells aren't executed in order or deleted after being run.
* For sessions in the built-in python console, we can use:
    ```python
    import readline
    with open(os.path.join(save_dir, 'config.py'), 'w') as f:
        history = [readline.get_history_item(i + 1) for i in range(readline.get_current_history_length())]
        f.write('\n'.join(history))
    ```

I think this will be pretty cool, I'll get started implementing it.